File: core/src/main/java/com/satis/viewmodel/core/BaseViewModel.java
package com.satis.viewmodel.core;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LifecycleOwner;
import androidx.lifecycle.Observer;

import java.util.HashMap;
import java.util.Map;

public class BaseViewModel extends AndroidViewModel {
    Map<String,BaseLiveData> mMutableLiveDataMap = new HashMap<>(4);
    public BaseViewModel(@NonNull Application application) {
        super(application);
    }

    public <T> void addObserver(@NonNull LifecycleOwner owner, String tag, BaseObserver<T> observer, boolean isSticky){
        BaseLiveData baseLiveData = mMutableLiveDataMap.get(tag);
        if (baseLiveData==null){
            baseLiveData = new BaseLiveData<T>();
            mMutableLiveDataMap.put(tag,baseLiveData);
        }
        baseLiveData.observe(owner, (Observer) observer, isSticky);
    }

    public void postValue(String tag,Object o){
        BaseLiveData baseLiveData = mMutableLiveDataMap.get(tag);
        if (baseLiveData!=null){
            baseLiveData.postValue(o);
        }
    }
    public void setValue(String tag,Object o){
        BaseLiveData baseLiveData = mMutableLiveDataMap.get(tag);
        if (baseLiveData!=null){
            baseLiveData.setValue(o);
        }
    }

}

File: viewmodel-core/src/main/java/com/satis/viewmodel/core/BaseViewModel.kt
package com.satis.viewmodel.core

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.Observer
import java.util.HashMap

open class BaseViewModel(application: Application) : AndroidViewModel(application) {
    private var mMutableLiveDataMap: MutableMap<String, BaseLiveData<Any?>> = HashMap(4)
    @Suppress("UNCHECKED_CAST")
    fun <T> addObserver(
        owner: LifecycleOwner,
        tag: String,
        observer: BaseObserver<T>,
        isSticky: Boolean
    ) {
        var baseLiveData = mMutableLiveDataMap[tag]
        if (baseLiveData == null) {
            baseLiveData = BaseLiveData<T>() as BaseLiveData<Any?>
            mMutableLiveDataMap[tag] = baseLiveData
        }
        baseLiveData.observe(owner, observer as Observer<Any?>, isSticky)
    }

    fun postValue(tag: String, o: Any?) {
        val baseLiveData = mMutableLiveDataMap[tag]
        baseLiveData?.postValue(o)
    }

    fun setValue(tag: String, o: Any?) {
        val baseLiveData = mMutableLiveDataMap[tag]
        baseLiveData?.setValue(o)
    }
}